src/lib/logical_query_plan/delete_node.hpp

```diff
@@ -12,11 +12,15 @@ namespace opossum {
  */
 class DeleteNode : public AbstractLQPNode {
  public:
+  static std::shared_ptr<DeleteNode> make(const std::string& table_name);
```

Comment:
another suggestion: The constructor parameters are duplicated here for every class type. Is this `::make()` our new way to create objects? I see the point, but it seems a bit cumbersome to implement all these method. We could simply create an alias for `std::make_shared<>` and call it `make<>`. It would be equally short and we didn’t have to manually implement it for each class.

```c++
DeleteNode::make(table_name);
make<DeleteNode>(table_name);  // same length
```

Reply:
I don't like the feeling of aliasing stdlib functions, but I agree that the `class::make` functions as they are right now are redundant. 
Currently I feel like making the child/children optional (i.e. defaulting to nullptr) parameters of the constructors. So 

`JoinNode::make(arg, left_child, right_child)`
would become
`std::make_shared<JoinNode>(arg, left_child, right_child)`

Reply:
This seems really complicated. What about adding a method `set_children` that returns the object?

```C++
std::make_shared<SomeNode>(args)->set_children(left_child, right_child);
```

Reply:
You were my only hope for someone to appreciate the template magic. :(

We had that discussion somewhere here: https://github.com/hyrise/hyrise/pull/673#discussion_r165935174

However, that "nested creation" thing is something that is limited to tests. Maybe we shouldn't touch the interface too much for it?

Reply:
Haha, sorry^^ It’s impressive for sure, but considering that it’s just a bit syntactic sugar, I wonder if it’s worth it.

Reply:
- @mrks that is cool, yet frightening, I wouldn't do it. 
- `set_children()` returning the node would work, but why not just...
- ...make the children normal constructor arguments, as I proposed? e.g. `JoinNode(JoinMode mode, ..., std::shared_ptr<AbstractLQPNode> left_child = nullptr,  std::shared_ptr<AbstractLQPNode> right_child = nullptr);`

This would make the code a bit longer, but still within the bounds of acceptability 

```c++
  const auto expected_optimized_lqp =
  std::make_shared<PredicateNode>(_mock_node_b, PredicateCondition::GreaterThan, 40,
    std::make_shared<PredicateNode>(_mock_node_b, PredicateCondition::GreaterThan, 50,
      std::make_shared<PredicateNode>(_mock_node_a, PredicateCondition::Equals, 42,
        ProjectionNode::make_forward_node(
          std::make_shared<PredicateNode>(_mock_node_c, PredicateCondition::LessThan, 500,
            std::make_shared<PredicateNode>(_mock_node_a, PredicateCondition::GreaterThanEquals, 90,
              _mock_node
  ))))));
```

Reply:
It would require all LQP nodes to deal with the children stuff. It's not perfect, but I could live with it.

Reply:
If you added  constructors to the base class accepting children, at least you wouldn’t have to call `set_left_child` and so on everywhere.

Reply:
Doing this right now, experiencing much hate for the fact that you can't pass initializer lists into std::make_shared

Reply:
This won't work. Crap. We can't wire up the LQP in the constructor because shared_from_this() is not callable in the constructor yet. So, back to <Node>::make? Or @mrks template magic solution? I give up for today. @mjendruk